zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1107 @@
+use std::rc::Rc;
+use zcash_client_backend::data_api::scanning::{ScanPriority, ScanRange};
+
+use incrementalmerkletree::{Address, Position};
+use zcash_primitives::consensus::{self, BlockHeight, NetworkUpgrade};
+
+use zcash_client_backend::data_api::SAPLING_SHARD_HEIGHT;
+
+use crate::error::SqliteClientError;
+use crate::{PRUNING_DEPTH, VALIDATION_DEPTH};
+
+use super::block_height_extrema;
+
+#[derive(Debug, Clone, Copy)]
+enum Insert {
+    Left,
+    Right,
+}
+
+impl Not for Insert {
+    type Output = Self;
+
+    fn not(self) -> Self::Output {
+        match self {
+            Insert::Left => Insert::Right,
+            Insert::Right => Insert::Left,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Dominance {
+    Left,
+    Right,
+    Equal,
+}
+
+impl From<Insert> for Dominance {
+    fn from(value: Insert) -> Self {
+        match value {
+            Insert::Left => Dominance::Left,
+            Insert::Right => Dominance::Right,
+        }
+    }
+}
+
+pub(crate) fn parse_priority_code(code: i64) -> Option<ScanPriority> {
+    use ScanPriority::*;
+    match code {
+        10 => Some(Scanned),
+        20 => Some(Historic),
+        30 => Some(OpenAdjacent),
+        40 => Some(FoundNote),
+        50 => Some(ChainTip),
+        60 => Some(Verify),
+        _ => None,
+    }
+}
+
+pub(crate) fn priority_code(priority: &ScanPriority) -> i64 {
+    use ScanPriority::*;
+    match priority {
+        Scanned => 10,
+        Historic => 20,
+        OpenAdjacent => 30,
+        FoundNote => 40,
+        ChainTip => 50,
+        Verify => 60,
+    }
+}
+
+pub(crate) fn suggest_scan_ranges(
+    conn: &rusqlite::Connection,
+    min_priority: Option<ScanPriority>,
+) -> Result<Vec<ScanRange>, SqliteClientError> {
+    let mut stmt_scan_ranges = conn.prepare_cached(
+        "SELECT block_range_start, block_range_end, priority
+         FROM scan_queue
+         WHERE priority >= :min_priority
+         ORDER BY priority DESC, block_range_end DESC",
+    )?;
+
+    let mut rows = stmt_scan_ranges.query(named_params![
+        ":min_priority": priority_code(&min_priority.unwrap_or(ScanPriority::Historic))
+    ])?;
+
+    let mut result = vec![];
+    while let Some(row) = rows.next()? {
+        let range = Range {
+            start: row.get::<_, u32>(0).map(BlockHeight::from)?,
+            end: row.get::<_, u32>(1).map(BlockHeight::from)?,
+        };
+        let code = row.get::<_, i64>(2)?;
+        let priority = parse_priority_code(code).ok_or_else(|| {
+            SqliteClientError::CorruptedData(format!("scan priority not recognized: {}", code))
+        })?;
+
+        result.push(ScanRange::from_parts(range, priority));
+    }
+
+    Ok(result)
+}
+
+// This implements the dominance rule for range priority. If the inserted range's priority is
+// `Verify`, this replaces any existing priority. Otherwise, if the current priority is
+// `Scanned`, this overwrites any priority
```

Comment:
```suggestion
// `Verify`, this replaces any existing priority. Otherwise, if the current priority is
// `Scanned`, it remains as `Scanned`; and if the new priority is `Scanned`, it
// overrides any existing priority.
```